core/src/main/java/com/graphhopper/util/GHUtility.java

```diff
@@ -317,32 +326,55 @@ public static Graph shuffle(Graph g, Graph sortedGraph) {
      */
     public static Graph sortDFS(Graph g, Graph sortedGraph) {
         int nodes = g.getNodes();
-        final GHIntArrayList list = new GHIntArrayList(nodes);
-        list.fill(nodes, -1);
-        final GHBitSetImpl bitset = new GHBitSetImpl(nodes);
-        final AtomicInteger ref = new AtomicInteger(-1);
+        final GHIntArrayList nodeList = new GHIntArrayList(nodes);
+        nodeList.fill(nodes, -1);
+        final GHBitSetImpl nodeBitset = new GHBitSetImpl(nodes);
+        final AtomicInteger nodeRef = new AtomicInteger(-1);
+
+        int edges = g.getEdges();
+        final GHIntArrayList edgeList = new GHIntArrayList(edges);
+        edgeList.fill(edges, -1);
+        final GHBitSetImpl edgeBitset = new GHBitSetImpl(edges);
+        final AtomicInteger edgeRef = new AtomicInteger(-1);
+
         EdgeExplorer explorer = g.createEdgeExplorer();
         for (int startNode = 0; startNode >= 0 && startNode < nodes;
-             startNode = bitset.nextClear(startNode + 1)) {
+             startNode = nodeBitset.nextClear(startNode + 1)) {
             new DepthFirstSearch() {
                 @Override
                 protected GHBitSet createBitSet() {
-                    return bitset;
+                    return nodeBitset;
+                }
+
+                @Override
+                protected boolean checkAdjacent(EdgeIteratorState edge) {
+                    int edgeId = edge.getEdge();
+                    if (!edgeBitset.contains(edgeId)) {
+                        edgeBitset.add(edgeId);
+                        edgeList.set(edgeRef.incrementAndGet(), edgeId);
+                    }
+                    return super.checkAdjacent(edge);
                 }
 
                 @Override
                 protected boolean goFurther(int nodeId) {
-                    list.set(nodeId, ref.incrementAndGet());
+                    nodeList.set(nodeId, nodeRef.incrementAndGet());
                     return super.goFurther(nodeId);
                 }
             }.start(explorer, startNode);
         }
-        return createSortedGraph(g, sortedGraph, list);
+        return createSortedGraph(g, sortedGraph, nodeList, edgeList);
     }
 
-    static Graph createSortedGraph(Graph fromGraph, Graph toSortedGraph, final IntIndexedContainer oldToNewNodeList) {
-        AllEdgesIterator eIter = fromGraph.getAllEdges();
-        while (eIter.next()) {
+    static Graph createSortedGraph(Graph fromGraph, Graph toSortedGraph, final IntIndexedContainer oldToNewNodeList, final IntIndexedContainer newToOldEdgeList) {
+        int edges = fromGraph.getEdges();
+        for (int i = 0; i < edges; i++) {
+            int edgeId = newToOldEdgeList.get(i);
+            if (edgeId < 0)
+                continue;
+
+            EdgeIteratorState eIter = fromGraph.getEdgeIteratorState(edgeId, Integer.MIN_VALUE);
```

Comment:
Did you try how long the sorting takes now (say for a continental sized graph) ? Usually calling `getEdgeIteratorState` many times is not very efficient, because it creates a new object on every call and thus leads to lots of GC.

This goes beyond this PR, but I think its odd that the sorting methods are not part of the `(Base)Graph` API. For example if a `Graph` has a `TurnCostExtension` we need to sort the turn cost entries also, but given a `Graph` I cannot even know what kind of `GraphExtension` it is (see #1398). Similarly, I have to remember to sort the node coordinates here. More generally, I should not need to know all these details about the graph to sort it, the sorting details should be handled within the graph implementation.

Reply:
No, I originally implemented this and successfully ran it on the world by adding a capability to `AllEdgesIterator` to accept an edge ordering provider (`newToOldEdgeIndex`), but that diff got a lot larger (https://github.com/graphhopper/graphhopper/pull/1757/files/f6c4354d243330b2b9126eba7ff9404a324c9669) whereas this change was more self-contained and I assumed it was equivalent.

I could go back to that more complex approach if you think it would be better? Or add an API to `AllEdgesIterator` to set the edgeId?

I agree moving this into `(Base)Graph` would be good, but would involve changing more parts of the code than I'm familiar with at this point so probably better for another PR.

Reply:
Ok, no I do not think the more complex approach is needed at the moment. I will probably move the sorting into the graph for #1398, so I guess we can leave this as is and optimize in another PR.

Reply:
OK, just tested this out sorting the world. There were slightly more GC activity during the sort operation compared to modifying `AllEdgesIterator`(30 GC events vs. 23 - both with 56G RAM and java 8 with G1GC) and both finished in <200 seconds which seems reasonable.

verbose GC log from modifying `AllEdgesIterator`:
```
[GC pause (G1 Evacuation Pause) (young) 26717M->25106M(57344M), 0.0923583 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 27602M->25378M(57344M), 0.0365206 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0005099 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0466707 secs]
[GC remark, 0.0493437 secs]
[GC cleanup 25418M->25418M(57344M), 0.0036449 secs]
[GC pause (G1 Evacuation Pause) (young) 27954M->25779M(57344M), 0.0380569 secs]
[GC pause (G1 Evacuation Pause) (young) 28723M->26229M(57344M), 0.0444583 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 30501M->26892M(57344M), 0.0583504 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0008904 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0477114 secs]
[GC remark, 0.0490855 secs]
[GC cleanup 26932M->26932M(57344M), 0.0034052 secs]
[GC pause (G1 Evacuation Pause) (young) 31964M->27676M(57344M), 0.0723461 secs]
[GC pause (G1 Evacuation Pause) (young) 33820M->28649M(57344M), 0.0871456 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 35721M->29789M(57344M), 0.1029310 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0020978 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0483938 secs]
[GC remark, 0.0548950 secs]
[GC cleanup 29845M->29845M(57344M), 0.0035427 secs]
[GC pause (G1 Evacuation Pause) (young) 37789M->31077M(57344M), 0.1149464 secs]
[GC pause (G1 Evacuation Pause) (young) 39861M->32503M(57344M), 0.1292224 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 42039M->33986M(57344M), 0.1374616 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0032296 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0495207 secs]
[GC remark, 0.0535295 secs]
[GC cleanup 34034M->34034M(57344M), 0.0029495 secs]
[GC pause (G1 Evacuation Pause) (young) 47064M->36126M(57344M), 0.1380340 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 44638M->37473M(57344M), 0.1547417 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0043262 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0528809 secs]
[GC remark, 0.0627294 secs]
[GC cleanup 37529M->37529M(57344M), 0.0033535 secs]
[GC pause (G1 Evacuation Pause) (young) 47537M->39121M(57344M), 0.1481099 secs]
[GC pause (G1 Evacuation Pause) (young) 47825M->40709M(57344M), 0.1504248 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 48101M->42066M(57344M), 0.1330208 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0060035 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0527521 secs]
[GC remark, 0.0726408 secs]
[GC cleanup 42114M->42114M(57344M), 0.0032893 secs]
[GC pause (G1 Evacuation Pause) (young) 48274M->43156M(57344M), 0.1108996 secs]
```

verbose GC log with `getEdgeIteratorState`:
```
[GC pause (G1 Evacuation Pause) (young) 24677M->24164M(57344M), 0.1245786 secs]
[GC pause (G1 Evacuation Pause) (young) 26660M->25057M(57344M), 0.0654925 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 27553M->25211M(57344M), 0.0306106 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0011887 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0586427 secs]
[GC remark, 0.0597587 secs]
[GC cleanup 25323M->25323M(57344M), 0.0030835 secs]
[GC pause (G1 Evacuation Pause) (young) 27915M->25500M(57344M), 0.0259677 secs]
[GC pause (G1 Evacuation Pause) (young) 27996M->25790M(57344M), 0.0356115 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 28590M->26119M(57344M), 0.0519495 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0012558 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0372878 secs]
[GC remark, 0.0478013 secs]
[GC cleanup 26192M->26192M(57344M), 0.0031108 secs]
[GC pause (G1 Evacuation Pause) (young) 29655M->26520M(57344M), 0.0723304 secs]
[GC pause (G1 Evacuation Pause) (young) 30664M->26972M(57344M), 0.0831304 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 32220M->27550M(57344M), 0.1013616 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0016264 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0378618 secs]
[GC remark, 0.0491968 secs]
[GC cleanup 27624M->27624M(57344M), 0.0028098 secs]
[GC pause (G1 Evacuation Pause) (young) 33774M->28257M(57344M), 0.1214522 secs]
[GC pause (G1 Evacuation Pause) (young) 34465M->28932M(57344M), 0.0785504 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 36532M->29837M(57344M), 0.1537344 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0017540 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0390381 secs]
[GC remark, 0.0511714 secs]
[GC cleanup 29917M->29917M(57344M), 0.0028430 secs]
[GC pause (G1 Evacuation Pause) (young) 37213M->30694M(57344M), 0.0923386 secs]
[GC pause (G1 Evacuation Pause) (young) 43895M->32168M(57344M), 0.0942218 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 40136M->33114M(57344M), 0.1283859 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0031646 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0401712 secs]
[GC remark, 0.0556655 secs]
[GC cleanup 33194M->33194M(57344M), 0.0032600 secs]
[GC pause (G1 Evacuation Pause) (young) 44682M->34665M(57344M), 0.1275467 secs]
[GC pause (G1 Evacuation Pause) (young) 43769M->35967M(57344M), 0.1432414 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 45903M->37070M(57344M), 0.1242383 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0043771 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0421382 secs]
[GC remark, 0.0631872 secs]
[GC cleanup 37158M->37158M(57344M), 0.0029976 secs]
[GC pause (G1 Evacuation Pause) (young) 47742M->38286M(57344M), 0.1206002 secs]
[GC pause (G1 Evacuation Pause) (young) 48158M->39485M(57344M), 0.1221284 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 48445M->40676M(57344M), 0.1215852 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0056251 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.0414679 secs]
[GC remark, 0.0657385 secs]
[GC cleanup 40772M->40772M(57344M), 0.0029433 secs]
[GC pause (G1 Evacuation Pause) (young) 48436M->41736M(57344M), 0.1135536 secs]
[GC pause (G1 Evacuation Pause) (young) 48568M->42550M(57344M), 0.0950881 secs]
```